src/main/kotlin/org/opensearch/replication/task/index/IndexReplicationTask.kt

```diff
@@ -163,76 +164,91 @@ open class IndexReplicationTask(id: Long, type: String, action: String, descript
+                                pollForMetadata(this)
                             }
-                            else -> {
-                                state
+                        }
+
+                        if (state !is MonitoringState) {
+                            // Tasks need to be started
+                            state
+                        } else {
+                            state = pollShardTaskStatus((followingTaskState as FollowingState).shardReplicationTasks)
+                            followingTaskState = startMissingShardTasks((followingTaskState as FollowingState).shardReplicationTasks)
+                            when (state) {
+                                is MonitoringState -> {
+                                    updateMetadata()
+                                }
+                                is FailedState -> {
+                                    // Try pausing first if we get Failed state. This returns failed state if pause failed
+                                    pauseReplication(state)
+                                }
+                                else -> {
+                                    state
+                                }
                             }
                         }
                     }
+                    ReplicationState.FAILED -> {
+                        assert(currentTaskState is FailedState)
+                        failReplication(currentTaskState as FailedState)
+                        currentTaskState
+                    }
+                    ReplicationState.COMPLETED -> {
+                        markAsCompleted()
+                        CompletedState
+                    }
                 }
-                ReplicationState.FAILED -> {
-                    assert(currentTaskState is FailedState)
-                    failReplication(currentTaskState as FailedState)
-                    currentTaskState
-                }
-                ReplicationState.COMPLETED -> {
-                    markAsCompleted()
-                    CompletedState
+                if (newState != currentTaskState) {
+                    currentTaskState = updateState(newState)
                 }
+                if (isCompleted) break
             }
-            if (newState != currentTaskState) {
-                currentTaskState = updateState(newState)
+            catch(e: OpenSearchException) {
+                val status = e.status().status
+                // Index replication task shouldn't exit before shard replication tasks
+                // As long as shard replication tasks doesn't encounter any errors, Index task
+                // should continue to poll and Any failure encoutered from shard task should
+                // invoke state transition and exit
+                if(status < 500 && status != RestStatus.TOO_MANY_REQUESTS.status) {
+                    log.error("Exiting index replication task", e)
```

Comment:
should we print stack trace here ?

Reply:
I think this version of logger should log the stack trace as well.

Reply:
No Saim we need to use  ` ${e.stackTraceToString()}")`

Reply:
Tested this on local - the above format is useful during any logging within the string passed to logger. 
The above version in code should log the stack trace.